pageserver/src/storage_sync.rs

```diff
@@ -129,1330 +114,587 @@
-                    warn!("No remote index entry for timeline {sync_id}, skipping deletion");
-                    return Ok(());
+                UploadOp::Delete(_) => {
+                    upload_queue.num_inprogress_deletions += 1;
                 }
-            },
-        }
-    };
-
-    let timeline_path = conf.timeline_path(&sync_id.timeline_id, &sync_id.tenant_id);
-    let new_index_part =
-        IndexPart::from_remote_timeline(&timeline_path, updated_remote_timeline)
-            .context("Failed to create an index part from the updated remote timeline")?;
-
-    debug!("Uploading remote index for the timeline");
-    REMOTE_INDEX_UPLOAD
-        .with_label_values(&[
-            &sync_id.tenant_id.to_string(),
-            &sync_id.timeline_id.to_string(),
-        ])
-        .inc();
-
-    upload_index_part(conf, storage, sync_id, new_index_part)
-        .await
-        .context("Failed to upload new index part")
-}
+                UploadOp::Barrier(sender) => {
+                    sender.send_replace(());
+                    continue;
+                }
+            };
+
+            // Assign unique ID to this task
+            upload_queue.task_counter += 1;
+            let task_id = upload_queue.task_counter;
+
+            // Add it to the in-progress map
+            let task = Arc::new(UploadTask {
+                task_id,
+                op: next_op,
+                retries: AtomicU32::new(0),
+            });
+            upload_queue
+                .inprogress_tasks
+                .insert(task.task_id, Arc::clone(&task));
+
+            // Spawn task to perform the task
+            let self_rc = Arc::clone(self);
+            task_mgr::spawn(
+                self.runtime.handle(),
+                TaskKind::RemoteUploadTask,
+                Some(self.tenant_id),
+                Some(self.timeline_id),
+                "remote upload",
+                false,
+                async move {
+                    self_rc.perform_upload_task(task).await;
+                    Ok(())
+                },
+            );
 
-async fn validate_task_retries(
-    current_attempt: u32,
-    max_sync_errors: NonZeroU32,
-) -> ControlFlow<(), ()> {
-    let max_sync_errors = max_sync_errors.get();
-    if current_attempt >= max_sync_errors {
-        return ControlFlow::Break(());
+            // Loop back to process next task
+        }
     }
 
-    exponential_backoff(current_attempt, 1.0, 30.0).await;
-    ControlFlow::Continue(())
-}
+    ///
+    /// Perform an upload task.
+    ///
+    /// The task is in the `inprogress_tasks` list. This function will try to
+    /// execute it, retrying forever. On successful completion, the task is
+    /// removed it from the `inprogress_tasks` list, and any next task(s) in the
+    /// queue that were waiting by the completion are launched.
+    ///
+    async fn perform_upload_task(self: &Arc<Self>, task: Arc<UploadTask>) {
```

Comment:
If feels that `delete` operation implies a different name, maybe `sync_task`?

Reply:
I suppose. Then again, we also call it an upload queue. I think it's OK to think of deletions as uploads in this context. With "Sync", you might expect it to include downloads. Let's leave it as it is for now; we can rename it later if we feel like it.